param_passing/tutorial.md

```diff
@@ -0,0 +1,561 @@
+
+When `element_id=""` then it assumed that the element to be added is a direct
+child of the original model.
+
+#### Add examples
+
+**Example 1**
+
+Given the original model has a `//link[@name="chassis"]`, the below example
+adds the `//visual[@name="camera_visual"]` element as a direct child of
+`//link[@name="chassis"]`:
+
+```xml
+<!-- In //experimental:params -->
+<visual element_id="chassis" name="camera_visual" action="add">
+  <pose>1 0 0 0 0 0</pose>
+  <geometry>
+    <box>
+      <size>0.05 0.05 0.05</size>
+    </box>
+  </geometry>
+</visual>
+```
+
+The following example is equivalent (but it is recommended to `add` the element
+following the structure above when the element requires the `@name` attribute):
+
+```xml
+<!-- //experimental:params -->
+<link element_id="chassis">
+  <visual name="camera_visual" action="add">
+    <pose>1 0 0 0 0 0</pose>
+    <geometry>
+      <box>
+        <size>0.05 0.05 0.05</size>
+      </box>
+    </geometry>
+  </visual>
+</link>
+```
+
+In the above example, the `add` action is specified in a direct child of where
+`@element_id` is declared.
+
+**Example 2**
+
+For adding elements that do not have the `@name` attribute
+(e.g., `//transparency`, `//pose`, `//inertial`, or `//axis`),
+specify the `@action` in the direct children of the `@element_id` element.
+
+```xml
+<!-- //experimental:params -->
+<link element_id="chassis">
+  <inertial action="add">
+    <mass>1</mass>
+    <inertia>
+      ...
+    </inertia>
+  </inertial>
+</link>
+```
+
+```xml
+<!-- //experimental:params -->
+<visual element_id="top::camera_visual">
+  <transparency action="add">0.5</transparency>
+  <pose relative_to="something" action="add">0 0 1 0 0 0</pose>
+</visual>
+```
+
+**Example 3**
+
+To add a new element as a direct child of the included/original model, use `element_id=""`:
+
+```xml
+<!-- //experimental:params -->
+<frame element_id="" name="some_frame" action="add">
+  <pose>0 1 0 0 0 0</pose>
+</frame>
+```
+
+This adds the `//frame[@name="some_frame"]` to the included `//model`
+(i.e., the output would be `//model/frame[@name="some_frame"]`).
+
+---
+
+### Modify action
+
+The `modify` action provides the ability to modify values and/or attributes of elements.
+This action only updates existing elements and does not add or remove them.
+It can update or add attributes to existing elements but can not remove them.
+
+To `modify` elements and/or attributes, the full element path to that element is needed with the new value(s).
+For example, if the original model has a `//visual[@name="visual"]/geometry/sphere[radius=0.5]`
+and modifying the `//radius` is desired, then under `//experimental:params`:
+
+```xml
+<visual element_id="path::to::visual" action="modify">
+  <geometry>
+    <sphere>
+      <radius custom:attr="foo">1.0</radius>
+    </sphere>
+  </geometry>
+</visual>
+```
+
+This modifies the original `//radius` from 0.5 to be 1.0 and adds a new attribute to the element.
```

Comment:
mention that the new attribute is `cusom:attr`?

Reply:
215e07b977d9033233462127d373fdbb6f19013a